crates/ruff_server/src/trace.rs

```diff
@@ -0,0 +1,151 @@
+use lsp_types::{notification::Notification, TraceValue};
+use serde::Deserialize;
+use std::sync::{Mutex, OnceLock};
+use tracing::{level_filters::LevelFilter, Event};
+use tracing_subscriber::{layer::SubscriberExt, Layer};
+
+use crate::server::ClientSender;
+
+static LOGGING_SENDER: OnceLock<ClientSender> = OnceLock::new();
+
+static TRACE_VALUE: Mutex<lsp_types::TraceValue> = Mutex::new(lsp_types::TraceValue::Off);
+
+pub(crate) fn set_trace_value(trace_value: TraceValue) {
+    let mut global_trace_value = TRACE_VALUE
+        .lock()
+        .expect("trace value mutex should be available");
+    *global_trace_value = trace_value;
+}
+
+pub(crate) fn init_tracing(sender: ClientSender) {
+    LOGGING_SENDER
+        .set(sender)
+        .expect("logging sender should only be initialized once");
+
+    // TODO(jane): Provide a way to set the log level
```

Comment:
I actually explored having the log level be set by the trace value (you can see it in the commit history). I ultimately decided against this because the trace value sent by the LSP is [only meant for `$/logTrace`, not `window/logMessage`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#traceValue).

Reply:
I've introduced a `logLevel` setting to client settings to fix this.

Reply:
Oh I didn't realize that difference when reviewing yesterday. What's the reason we don't use `$/logTrace` because it seems like the right fit from the spec

> $/logTrace should be used for systematic trace reporting. For single debugging messages, the server should send [window/logMessage](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_logMessage) notifications. [source](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#logTrace)

Reply:
Okay, I think I'm starting to understand what this PR does. I think the PR title needs updating. 

I try to summarize my understanding to make sure it is correct. 

This PR only logs `trace::debug`, `trace::warn` but it omits all `trace::debug_span` etc. Is this correct?

What's the motivation for making this distinction instead of just logging all tracing events with `$/logTrace` for now. That seems simpler because it doesn't require additional filtering or yet another setting (which I'm worried that it just gets hard to understand for users, especially because `TRACE_VALUE: message` exists). Is it because you need a way to format a span to a string?

Reply:
@MichaReiser My plan was for tracing events to go through `window/logMessage` and then later implement a additional tracing layer that sends all tracing events / messages to `$/logTrace` as a follow-up.

The reason I'm doing this is to separate the log level from the trace level.